kedro/io/data_catalog.py

```diff
@@ -322,20 +324,17 @@ def _is_pattern(pattern: str):
         return "{" in pattern
 
     @staticmethod
-    def _match_pattern(
-        data_set_patterns: dict[str, dict[str, Any]], data_set_name: str
-    ) -> str | None:
+    def _match_pattern(data_set_patterns: Patterns, data_set_name: str) -> str | None:
         """Match a dataset name against patterns in a dictionary containing patterns"""
-        for pattern, _ in data_set_patterns.items():
-            result = parse(pattern, data_set_name)
-            if result:
-                return pattern
-        return None
+        matches = [
+            pattern
+            for pattern in data_set_patterns.keys()
+            if parse(pattern, data_set_name)
+        ]
+        return next(filter(None, matches), None)
```

Comment:
It's better to make it a generator, so we only parse until we find a match. Also the second filter is not needed here anymore because of the filter you've added in the generator.
```suggestion
        matches = (
            pattern
            for pattern in data_set_patterns.keys()
            if parse(pattern, data_set_name)
        )
        return next(matches, None)
```